### PR TITLE
Add pio_reader Fortran unit tests

### PR DESCRIPTION
**Pull request by nusbaume**
Tag name (required for release branches):
Originator(s): @mwaxmonsky @nusbaume

Description (include the issue title, and the keyword ['closes', 'fixes', 'resolves'] followed by the issue number):

This PR introduces pFUnit Fortran unit tests for the `pio_reader` implementation of the `ccpp_io_reader` abstract class that is used in CAM-SIMA, which should now run automatically with any PR to CAM-SIMA as part of the Fortran unit tests Github Action.

Note that although several tests are added by this PR, the code coverage is not complete, as several of the edge cases require specialized NetCDF files that will need to be created.  Github issue #404 can be used to keep track of this.

Describe any changes made to build system:

A       src/physics/utils/CMakeLists.txt
M       test/unit/fortran/CMakeLists.txt
A       test/unit/fortran/ccpp_framework_stub/CMakeLists.txt
A       test/unit/fortran/cmake/FindNetCDF.cmake
A       test/unit/fortran/cmake/FindPIO.cmake
A       test/unit/fortran/mock_routines/CMakeLists.txt
A       test/unit/fortran/src/pio_reader/CMakeLists.tx
  - Add files needed to build the pio_reader unit tests and their dependencies using CMake.

Describe any changes made to the namelist: N/A

List any changes to the defaults for the input datasets (e.g. boundary datasets): N/A

List all files eliminated and why: N/A

List all files added and what they do:

A       test/unit/fortran/ccpp_framework_stub/ccpp_kinds.F90
A       test/unit/fortran/mock_routines/cam_pio_utils.F90
A       test/unit/fortran/src/pio_reader/test_pio_reader.pf
  - Add pio_reader unit tests (and associated dependencies).

List all existing files that have been modified, and describe the changes: 
(Helpful git command: `git diff --name-status development...<your_branch_name>`)

M       .github/workflows/fortran_unit_tests.yml
  - Add pio_reader unit tests to Github Action worfklow.

M       .gitignore
  - Have git ignore all compiled Fortran files (".o" and ".mod").

M       src/physics/utils/pio_reader.F90
  - Fix bugs and remove un-needed host model dependencies  found during testing.

M       src/utils/cam_pio_utils.F90
  - Add optional "errcode" variable that can be used during testing (and to ensure that pio_reader doesn't kill the host model).

If there are new failures (compared to the `test/existing-test-failures.txt` file),
have them OK'd by the gatekeeper, note them here, and add them to the file.
If there are baseline differences, include the test and the reason for the
diff. What is the nature of the change? Roundoff?

derecho/intel/aux_sima:  ALL PASS

derecho/gnu/aux_sima:  ALL PASS

If this changes climate describe any run(s) done to evaluate the new
climate in enough detail that it(they) could be reproduced:

CAM-SIMA date used for the baseline comparison tests if different than latest:

### CODE REVIEW

File: test/unit/fortran/src/pio_reader/test_pio_reader.pf [hunk 1]
```diff
@@ -0,0 +1,1247 @@
...
+    ! Begin test:
+
+    reader = create_netcdf_reader_t()
+
+    ! Attempt to read variable without opening file:
+    call reader%get_var("pressure", pressure, errmsg, errcode)
+
+    ! Check that the correct error was raised:
+    @assertTrue(errcode /= 0)
+    @assertTrue(trim(errmsg) == expected_err_msg)
```
Comment (mwaxmonsky) [inline test/unit/fortran/src/pio_reader/test_pio_reader.pf:797]:
```suggestion
    @assertNotEqual(0, errcode)
    @assertEqual(expected_err_msg, trim(errmsg))
```

File: test/unit/fortran/src/pio_reader/test_pio_reader.pf [hunk 2]
```diff
@@ -0,0 +1,1247 @@
...
+    ! Begin test:
+
+    reader = create_netcdf_reader_t()
+
+    ! Attempt to read variable without opening file:
+    call reader%get_var("gas_names", gas_names, errmsg, errcode)
+
+    ! Check that the correct error was raised:
+    @assertTrue(errcode /= 0)
+    @assertTrue(trim(errmsg) == expected_err_msg)
```
Comment (mwaxmonsky) [inline test/unit/fortran/src/pio_reader/test_pio_reader.pf:838]:
```suggestion
    @assertNotEqual(0, errcode)
    @assertEqual(expected_err_msg, trim(errmsg))
```

File: test/unit/fortran/src/pio_reader/test_pio_reader.pf [hunk 3]
```diff
@@ -0,0 +1,1247 @@
...
+    ! Begin test:
+
+    reader = create_netcdf_reader_t()
+
+    ! Attempt to read variable without opening file:
+    call reader%get_var("press_ref", press_ref, errmsg, errcode)
+
+    ! Check that the correct error was raised:
+    @assertTrue(errcode /= 0)
+    @assertTrue(trim(errmsg) == expected_err_msg)
```
Comment (mwaxmonsky) [inline test/unit/fortran/src/pio_reader/test_pio_reader.pf:880]:
```suggestion
    @assertNotEqual(0, errcode)
    @assertEqual(expected_err_msg, trim(errmsg))
```

File: test/unit/fortran/src/pio_reader/test_pio_reader.pf [hunk 4]
```diff
@@ -0,0 +1,1247 @@
...
+
+    ! Close file:
+    call reader%close_file(errmsg, errcode)
+
+    ! Attempt to read variable after closing file:
+    call reader%get_var("pressure", pressure, errmsg, errcode)
+
+    ! Check that the correct error was raised:
+    @assertTrue(errcode /= 0)
+    @assertTrue(trim(errmsg) == expected_err_msg)
```
Comment (mwaxmonsky) [inline test/unit/fortran/src/pio_reader/test_pio_reader.pf:933]:
```suggestion
    @assertNotEqual(0, errcode)
    @assertEqual(expected_err_msg, trim(errmsg))
```

File: test/unit/fortran/src/pio_reader/test_pio_reader.pf [hunk 5]
```diff
@@ -0,0 +1,1247 @@
...
+
+    ! Open file:
+    call reader%open_file(fname, errmsg, errcode)
+
+    ! Attempt to read non-existent variable:
+    call reader%get_var("missing_var", missing_var, errmsg, errcode)
+
+    ! Check that the correct error was raised:
+    @assertTrue(errcode /= 0)
+    @assertTrue(trim(errmsg) == expected_err_msg)
```
Comment (mwaxmonsky) [inline test/unit/fortran/src/pio_reader/test_pio_reader.pf:979]:
```suggestion
    @assertNotEqual(0, errcode)
    @assertEqual(expected_err_msg, trim(errmsg))
```

File: test/unit/fortran/src/pio_reader/test_pio_reader.pf [hunk 6]
```diff
@@ -0,0 +1,1247 @@
...
```
Comment (mwaxmonsky) [inline test/unit/fortran/src/pio_reader/test_pio_reader.pf:1027]:
```suggestion
    @assertNotEqual(0, errcode)
    @assertEqual(expected_err_msg, trim(errmsg))
```

File: test/unit/fortran/src/pio_reader/test_pio_reader.pf [hunk 7]
```diff
@@ -0,0 +1,1247 @@
...
```
Comment (mwaxmonsky) [inline test/unit/fortran/src/pio_reader/test_pio_reader.pf:1077]:
```suggestion
    @assertNotEqual(0, errcode)
    @assertEqual(expected_err_msg, trim(errmsg))
```

File: test/unit/fortran/src/pio_reader/test_pio_reader.pf [hunk 8]
```diff
@@ -0,0 +1,1247 @@
...
+
+    ! Open file:
+    call reader%open_file(fname, errmsg, errcode)
+
+    ! Attempt to read variable with bad rank:
+    call reader%get_var("pressure", pressure, errmsg, errcode)
+
+    ! Check that the correct error was raised:
+    @assertTrue(errcode /= 0)
+    @assertTrue(trim(errmsg) == expected_err_msg)
```
Comment (mwaxmonsky) [inline test/unit/fortran/src/pio_reader/test_pio_reader.pf:1125]:
```suggestion
    @assertNotEqual(0, errcode)
    @assertEqual(expected_err_msg, trim(errmsg))
```

File: test/unit/fortran/src/pio_reader/test_pio_reader.pf [hunk 9]
```diff
@@ -0,0 +1,1247 @@
...
+
+    ! Open file:
+    call reader%open_file(fname, errmsg, errcode)
+
+    ! Attempt to read variable with bad rank:
+    call reader%get_var("gas_names", gas_names, errmsg, errcode)
+
+    ! Check that the correct error was raised:
+    @assertTrue(errcode /= 0)
+    @assertTrue(trim(errmsg) == expected_err_msg)
```
Comment (mwaxmonsky) [inline test/unit/fortran/src/pio_reader/test_pio_reader.pf:1173]:
```suggestion
    @assertNotEqual(0, errcode)
    @assertEqual(expected_err_msg, trim(errmsg))
```

File: test/unit/fortran/src/pio_reader/test_pio_reader.pf [hunk 10]
```diff
@@ -0,0 +1,1247 @@
...
+
+    ! Open file:
+    call reader%open_file(fname, errmsg, errcode)
+
+    ! Attempt to read variable with bad rank:
+    call reader%get_var("press_ref", press_ref, errmsg, errcode)
+
+    ! Check that the correct error was raised:
+    @assertTrue(errcode /= 0)
+    @assertTrue(trim(errmsg) == expected_err_msg)
```
Comment (mwaxmonsky) [inline test/unit/fortran/src/pio_reader/test_pio_reader.pf:1223]:
```suggestion
    @assertNotEqual(0, errcode)
    @assertEqual(expected_err_msg, trim(errmsg))
```

File: test/unit/fortran/src/pio_reader/test_pio_reader.pf [hunk 11]
```diff
@@ -0,0 +1,1247 @@
...
+
+    use mpi, only: mpi_finalize
+
+    integer :: mpi_ierr
+
+    ! Shutdown MPI:
+    ! NOTE: This must be done for the last test in the test suite.
+    call mpi_finalize(mpi_ierr)
+
+    @assertTrue(.true.)
```
Comment (mwaxmonsky) [inline test/unit/fortran/src/pio_reader/test_pio_reader.pf:1245]:
```suggestion
    use mpi, only: mpi_finalize, MPI_SUCCESS

    integer :: mpi_ierr

    ! Shutdown MPI:
    ! NOTE: This must be done for the last test in the test suite.
    call mpi_finalize(mpi_ierr)

    @assertEqual(MPI_SUCCESS, mpi_ierr)
```
This might not be necessary since we're just trying to be nice but maybe we do because another MPI test might run after this?

File: test/unit/fortran/src/pio_reader/test_pio_reader.pf [hunk 12]
```diff
@@ -0,0 +1,1247 @@
...
+
+end subroutine final_test
```
Comment (mwaxmonsky) [inline test/unit/fortran/src/pio_reader/test_pio_reader.pf:1247]:
```suggestion
end subroutine final_test

```

File: test/unit/fortran/CMakeLists.txt [hunk 13]
```diff
@@ -14,7 +15,7 @@ endif()
 
 # Overwrite the init flags chosen by CMake
 if(${CMAKE_Fortran_COMPILER_ID} MATCHES "GNU")
-  add_compile_options("-ffree-line-length-none")
+  add_compile_options(-ffree-line-length-none -std=f2018)
```
[the remark this answers is not in the document]
Comment (nusbaume) [reply test/unit/fortran/CMakeLists.txt]: Nope, good catch!  The `std` flag has been removed.

File: test/unit/fortran/CMakeLists.txt [hunk 14]
```diff
@@ -24,9 +25,25 @@ endif()
...
   find_package(PFUNIT REQUIRED)
   enable_testing()
-  add_subdirectory(src/core_utils)
+
+  if(CAM_SIMA_ENABLE_TESTS)
+    add_subdirectory(src/core_utils)
+  endif()
+  if(CAM_SIMA_ENABLE_IO_TESTS)
+    find_package(MPI REQUIRED COMPONENTS Fortran C)
+    list(APPEND CMAKE_MODULE_PATH "${CMAKE_CURRENT_SOURCE_DIR}/cmake")
```
[the remark this answers is not in the document]
Comment (nusbaume) [reply test/unit/fortran/CMakeLists.txt]: Makes sense to me!  The `find_package` calls should now be grouped together.

File: test/unit/fortran/src/pio_reader/test_pio_reader.pf [hunk 15]
```diff
@@ -0,0 +1,1247 @@
...
+
+    ! Attempt to open file:
+    call reader%open_file(fname, errmsg, errcode)
+
+    ! Attempt to open file again:
+    call reader%open_file(fname, errmsg, errcode)
+
+    ! Check that the correct error was raised:
+    @assertTrue(errcode /= 0)
+    @assertTrue(trim(errmsg) == expected_err_msg)
```
[the remark this answers is not in the document]
Comment (nusbaume) [reply test/unit/fortran/src/pio_reader/test_pio_reader.pf]: Done!

File: test/unit/fortran/src/pio_reader/test_pio_reader.pf [hunk 16]
```diff
@@ -0,0 +1,1247 @@
...
+    ! Check that the variable was read successfully:
+    @assertTrue(errcode == 0)
+    @assertTrue(errmsg == '')
+
+    ! Check that the variable's properties are correct:
+    @assertTrue(associated(vmr_ref))
+    ! Change
+    @assertTrue(size(vmr_ref, dim=1) == 2)
+    @assertTrue(size(vmr_ref, dim=2) == 20)
+    @assertTrue(size(vmr_ref, dim=3) == 14)
```
[the remark this answers is not in the document]
Comment (nusbaume) [reply test/unit/fortran/src/pio_reader/test_pio_reader.pf]: Done!

File: test/unit/fortran/src/pio_reader/test_pio_reader.pf (hunk 3 above)
Comment (nusbaume) [reply test/unit/fortran/src/pio_reader/test_pio_reader.pf]: Done!

File: test/unit/fortran/src/pio_reader/test_pio_reader.pf (hunk 4 above)
Comment (nusbaume) [reply test/unit/fortran/src/pio_reader/test_pio_reader.pf]: Done!

File: test/unit/fortran/src/pio_reader/test_pio_reader.pf (hunk 8 above)
Comment (nusbaume) [reply test/unit/fortran/src/pio_reader/test_pio_reader.pf]: Done!

File: test/unit/fortran/src/pio_reader/test_pio_reader.pf (hunk 9 above)
Comment (nusbaume) [reply test/unit/fortran/src/pio_reader/test_pio_reader.pf]: Done!

File: test/unit/fortran/src/pio_reader/test_pio_reader.pf (hunk 10 above)
Comment (nusbaume) [reply test/unit/fortran/src/pio_reader/test_pio_reader.pf]: Done!

File: test/unit/fortran/src/pio_reader/test_pio_reader.pf (hunk 12 above)
Comment (nusbaume) [reply test/unit/fortran/src/pio_reader/test_pio_reader.pf]: Done!

File: test/unit/fortran/src/pio_reader/test_pio_reader.pf [hunk 17]
```diff
@@ -0,0 +1,1247 @@
...
+
+    use ccpp_io_reader, only: abstract_netcdf_reader_t
+    use ccpp_io_reader, only: create_netcdf_reader_t
+
+    class(abstract_netcdf_reader_t), allocatable :: reader
+
+    integer            :: errcode
+    character(len=256) :: errmsg
+
+    integer            :: mpi_ierr
```
[the remark this answers is not in the document]
Comment (nusbaume) [reply test/unit/fortran/src/pio_reader/test_pio_reader.pf]: Good catch!  Removed the extraneous `mpi_ierr` variable.

File: .github/workflows/fortran_unit_tests.yml [hunk 18]
```diff
@@ -34,16 +38,45 @@ jobs:
...
+            -DCMAKE_PREFIX_PATH="$GITHUB_WORKSPACE/pFUnit/build/installed;/usr/lib/x86_64-linux-gnu/cmake/netCDF;/usr/lib/x86_64-linux-gnu/" \
             -DCAM_SIMA_ENABLE_CODE_COVERAGE=ON \
+            -DCAM_SIMA_ENABLE_IO_TESTS=ON \
+            -DCAM_SIMA_ENABLE_TESTS=ON \
             -B./build \
             -S./test/unit/fortran
           cd build
           make
 
+      - name: Checkout datafiles
```
[the remark this answers is not in the document]
Comment (nusbaume) [reply .github/workflows/fortran_unit_tests.yml]: Done!

File: src/physics/utils/pio_reader.F90 [hunk 19]
```diff
@@ -106,11 +106,12 @@ subroutine close_netcdf_file(this, errmsg, errcode)
       end if
 
       !Close NetCDF File:
-      call pio_closefile(this%sima_pio_fh%pio_fh)
+      call cam_pio_closefile(this%sima_pio_fh%pio_fh)
 
-      !Inidcate that file handle array id is no longer in use:
+      !Inidcate that file handle array id is no longer in use,
```
[the remark this answers is not in the document]
Comment (nusbaume) [reply src/physics/utils/pio_reader.F90]: Good catch!  Fixed!
